opendevin/events/serialization/action.py

```diff
@@ -58,6 +60,8 @@ def action_from_dict(action: dict) -> Action:
     args = action.get('args', {})
     try:
         decoded_action = action_class(**args)
+        decoded_action._source = action.get('source', EventSource.USER)
+        logger.info(decoded_action, extra={'msg_type': 'ACTION'})
```

Comment:
This part of the code is called on text coming from the LLM, so if we have to set the source here (do we have to?), I think it should be AGENT. Why do you think it should be USER?

Related, I didn't think _source is supposed to be sent to the LLM, nor returned. I'm a little curious, in what case did this `action.get('source',...)` work for you to get the source, not the default?

Reply:
+1, why manually set source?

Reply:
Actions from UI to Agent like MessageAction. Action from LLM is going to _step method https://github.com/OpenDevin/OpenDevin/blob/c4ad3b8f479e32e449d20a31797dd2ce379ab48a/opendevin/controller/agent_controller.py#L309

Where the source for User MessageAction is set?